### DISCUSSION TITLE
Skip to next (text) occurrence

### DISCUSSION THREAD

**Discussion by TontonSancho** [godotengine/godot-proposals#9003] [Editor]
It could be usefull to add a `ui_text_skip_selection_for_next_occurrence` implementation to Godot Editor.
Shortcut proposal this new action : `Ctrl+Alt+D` (as it still free)

As a daily user of Sublime-text, I'm enjoying the `Ctrl+D` implementation in Godot Editor.
`Ctrl+D` editor shortcut stands for : `ui_text_add_selection_for_next_occurrence`
What's missing (IMHO) is the ability to skip a selection occurrence.

To illustrate :
```
var foo:int
foo = 8
foo = footer_processor(foo)
return foo
```
In the previous example, I usually select the first `foo`, then press `Ctrl+D` several times to select all foos.
On Sublime Text, when `Ctrl+D` take me on the `foo` of `footer_precessor`, I'm pressing `Ctrl+K+N` to skip this occurrence (this goes on the next one and select it, like `Ctrl+D` would have done).

--
Have a great (godot editing) day

**Comment by TontonSancho** [comment]
n.b.

If someone already knows where shortcuts and text editor codes related stuffs are in the sources, I can manage to implement it.

--
Best regards

**Reply by Calinou** [reply]
The code that does this is here: https://github.com/godotengine/godot/blob/9adb7c7d130c6d4eb0e80b92d6eebd71fac3384d/scene/gui/text_edit.cpp#L5065-L5102

The shortcut name is `ui_text_add_selection_for_next_occurrence`.

**Reply by TontonSancho** [reply]
Thanks for the pointer.

I managed to find where to add stuffs :
- [text_edit.cpp](https://github.com/godotengine/godot/blob/9adb7c7d130c6d4eb0e80b92d6eebd71fac3384d/scene/gui/text_edit.cpp)
- [text_edit.h](https://github.com/godotengine/godot/blob/9adb7c7d130c6d4eb0e80b92d6eebd71fac3384d/scene/gui/text_edit.h)
- [input_map.cpp](https://github.com/godotengine/godot/blob/9adb7c7d130c6d4eb0e80b92d6eebd71fac3384d/core/input/input_map.cpp#L357) and [L694](https://github.com/godotengine/godot/blob/9adb7c7d130c6d4eb0e80b92d6eebd71fac3384d/core/input/input_map.cpp#L694)
- [ProjectSettings.xml](https://github.com/godotengine/godot/blob/9adb7c7d130c6d4eb0e80b92d6eebd71fac3384d/doc/classes/ProjectSettings.xml#L1248)

Do I forgot something ?

**Reply by Calinou** [reply]
Yes, these are the files you'll need to change to implement this :slightly_smiling_face:

**Comment by Calinou** [comment]
This exists in VS Code as <kbd>Ctrl + K</kbd> then <kbd>Ctrl + D</kbd>. It acts differently though, since it *moves* your selection to the next word.

**Reply by TontonSancho** [reply]
:thinking: Hum ... interesting side note.

Many question then :
- Is there an existing godot editor shortcut doing similar VS Code behavior (I mean, moving selection to next word) ? (currently, `grep` did not help me to find one)
- Is there a shortcut chaining mechanism in godot editor to support such a combo-shortcuts ?
- What shortcut are you suggesting for adding the feature to godot editor ? (Personally, I don't mind)

**Reply by Calinou** [reply]
> Is there a shortcut chaining mechanism in godot editor to support such a combo-shortcuts ?

No, and I don't find this particular shortcut combination very usable in practice. I prefer the idea of using <kbd>Ctrl + Alt + D</kbd> instead.

**Reply by TontonSancho** [reply]
I will take a look at implementing this feature with the VS Code behavior in mind, I mean, `Ctrl + Alt + D` will move the selection to the next occurrence at first goal but with support multi-selection state if any.

**Comment by TontonSancho** [comment]
[PR](https://github.com/godotengine/godot/pull/87883) added.

Do I have to close this discussion ?
